File: core/inpainter/state.py
from __future__ import annotations

import json
import os
from pathlib import Path

from inpainter.paths import core_state_dir

STATE_FILE = "state.json"


def load_state() -> str | None:
    path = _state_path()
    if not path.exists():
        return None
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None
    if not isinstance(data, dict):
        return None
    state = data.get("state")
    return state if isinstance(state, str) else None


def persist_state(state: str) -> None:
    directory = core_state_dir()
    directory.mkdir(parents=True, exist_ok=True)
    if os.name != "nt":
        directory.chmod(0o700)
    path = directory / STATE_FILE
    path.write_text(json.dumps({"state": state}, separators=(",", ":")), encoding="utf-8")
    if os.name != "nt":
        path.chmod(0o600)


def _state_path() -> Path:
    return core_state_dir() / STATE_FILE

File: core/src/state.ts
import { chmodSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";

import { coreStateDir } from "./paths.ts";

const STATE_FILE = "state.json";

export function loadState(): string | null {
  try {
    const data = JSON.parse(readFileSync(statePath(), "utf8")) as unknown;
    if (data === null || typeof data !== "object" || Array.isArray(data)) {
      return null;
    }
    const state = (data as { state?: unknown }).state;
    return typeof state === "string" ? state : null;
  } catch {
    return null;
  }
}

export function persistState(state: string): void {
  const directory = coreStateDir();
  mkdirSync(directory, { recursive: true });
  restrictDir(directory);
  const path = join(directory, STATE_FILE);
  writeFileSync(path, JSON.stringify({ state }));
  restrictFile(path);
}

function statePath(): string {
  return join(coreStateDir(), STATE_FILE);
}

function restrictDir(path: string): void {
  if (process.platform !== "win32") {
    chmodSync(path, 0o700);
  }
}

function restrictFile(path: string): void {
  if (process.platform !== "win32") {
    chmodSync(path, 0o600);
  }
}
